File: Classes/ThemeObject/CALayer+LWTheme.m
//
//  CALayer+LWTheme.m
//  LWThemeObjC
//
//  Created by lalawue on 2019/12/21.
//

#import "CALayer+LWTheme.h"
#import "LWTheme_Tools.h"

@implementation CALayer (LWTheme)

- (NSString *)lw_backgroundColor
{
    return (NSString *)[self lw_valueForSetter:@selector(setLw_backgroundColor:)];
}

- (void)setLw_backgroundColor:(NSString *)lw_backgroundColor
{
    UIColor *colorValue = [self lw_colorForPattern:lw_backgroundColor];
    self.backgroundColor = colorValue.CGColor;
    [self lw_storeValue:lw_backgroundColor forSetter:@selector(setLw_backgroundColor:)];
}

- (NSString *)lw_borderColor
{
    return (NSString *)[self lw_valueForSetter:@selector(setLw_borderColor:)];
}

- (void)setLw_borderColor:(NSString *)lw_borderColor
{
    UIColor *colorValue = [self lw_colorForPattern:lw_borderColor];
    self.borderColor = colorValue.CGColor;
    [self lw_storeValue:lw_borderColor forSetter:@selector(setLw_borderColor:)];
}

- (NSString *)lw_shadowColor
{
    return (NSString *)[self lw_valueForSetter:@selector(setLw_shadowColor:)];
}

- (void)setLw_shadowColor:(NSString *)lw_shadowColor
{
    UIColor *colorValue = [self lw_colorForPattern:lw_shadowColor];
    self.shadowColor = colorValue.CGColor;
    [self lw_storeValue:lw_shadowColor forSetter:@selector(setLw_shadowColor:)];
}

@end

File: Classes/ThemeObject/CALayer+LWTheme.swift
//
//  CALayer+LWTheme.swift
//  LWThemeSwift
//
//  Created by lalawue on 2019/12/22.
//

import UIKit

extension CALayer {
    
    @objc public var lw_backgroundColor: String? {
        
        get {
            return lw_valueForSetter(#selector(setter: CALayer.lw_backgroundColor)) as? String
        }
        
        set (colorPattern) {
            self.backgroundColor = lw_colorForPattern(colorPattern)?.cgColor
            lw_storeValue(colorPattern, setter: #selector(setter: CALayer.lw_backgroundColor))
        }
    }
    
    @objc public var lw_borderColor: String? {
        
        get {
            return lw_valueForSetter(#selector(setter: CALayer.lw_borderColor)) as? String
        }
        
        set (colorPattern) {
            self.borderColor = lw_colorForPattern(colorPattern)?.cgColor
            lw_storeValue(colorPattern, setter: #selector(setter: CALayer.lw_borderColor))
        }
    }
    
    @objc public var lw_shadowColor: String? {
        
        get {
            return lw_valueForSetter(#selector(setter: CALayer.lw_shadowColor)) as? String
        }
        
        set (colorPattern) {
            self.shadowColor = lw_colorForPattern(colorPattern)?.cgColor
            lw_storeValue(colorPattern, setter: #selector(setter: CALayer.lw_shadowColor))
        }
    }
}
